### PR TITLE
chore(dtfs2-6995): remove moment from azure-functions

### PR DESCRIPTION
**Pull request by MarRobSoftwire**
## Introduction :pencil2:
Moment js is deprecated, it should be replaced.

## Resolution :heavy_check_mark:
Replace moment() with vanilla Dates where possible and date-fns where necessary in `azure-functions/acbs-functions`. #2769 established tests for the date helpers which are aimed to be matched within reason, but some tests have been changed

### Changes to tests:
- removed `invalidDateFormats` - these were all 2 year dates (e.g. yy/MM/dd) & will be parsed (incorrectly) but there's no reason why a date `24/03/02` should be passed into these functions
- removed tests that epochs stored as strings shouldn't be parsed - these should never be passed into the functions & given the broad range of strings that are parsed, I see no reason to validate against this
- removed adding `true` day/month/years - this should never happen but I was testing any edge case I could think of
- removed tests that scientific notation shouldn't be parsed - should never be passed in so no reason to validate against
- removed tests for the maximum date value - no need to worry about deals being created in the year 275760

## Misc ➕ 
- remove `formatTimestamp` helper as it is identical to formatDate`
- #2769 missed moment usage in `get-next-quarter-date` so added a helper & test coverage

### CODE REVIEW

File: azure-functions/acbs-function/helpers/date.is-epoch.test.js [hunk 1]
```diff
@@ -72,16 +67,6 @@ describe('isEpoch', () => {
       mockValue: '8640000000000000', // https://262.ecma-international.org/5.1/#sec-15.9.1.1
       expected: true,
     },
-    {
```
Comment (MarRobSoftwire) [inline azure-functions/acbs-function/helpers/date.is-epoch.test.js]: These tests were passed by the moment implementation but I think making them work is a waste of time

File: azure-functions/acbs-function/helpers/date.get-year-and-mmdd.test.js
Comment (MarRobSoftwire) [inline azure-functions/acbs-function/helpers/date.get-year-and-mmdd.test.js:1]: git thinks this is a change of file name but the new file is brand new - ignore the diff!

File: azure-functions/acbs-function/helpers/date.js [hunk 2]
```diff
@@ -69,10 +91,11 @@ const formatTimestamp = (dateStr) => moment(isDate(dateStr) || isString(dateStr)
...
  */
-// This function is never actually used
-const addDay = (date, day) => moment(date).add({ day }).format('YYYY-MM-DD');
+const formatDate = (dateStr) => formatOrReturnInvalidDate(
+  getDateFromStringOrNumber(dateStr),
+);
+
 /**
  * @param {string | number | Date} date as a date string, epoch time or Date object
  * @param {number} day number of months to add
```
Comment (AlexBramhill) [inline azure-functions/acbs-function/helpers/date.js:101]: update

File: azure-functions/acbs-function/helpers/date.js [hunk 3]
```diff
@@ -89,7 +112,11 @@ const addDay = (date, day) => moment(date).add({ day }).format('YYYY-MM-DD');
...
  */
-const addMonth = (date, months) => moment(date).add({ months }).format('YYYY-MM-DD');
+const addMonth = (date, months) => {
+  const parsedDate = date instanceof Date ? date : getDateFromStringOrNumber(date);
+
+  return formatOrReturnInvalidDate(add(parsedDate, { months }));
+};
 /**
  * @param {string | number | Date} date as a date string, epoch time or Date object
  * @param {number} day number of months to add
```
Comment (AlexBramhill) [inline azure-functions/acbs-function/helpers/date.js:122]: update

File: azure-functions/acbs-function/helpers/date.format-date.test.js [hunk 4]
```diff
@@ -67,7 +60,7 @@ describe('formatDate', () => {
     },
     {
       description: 'should parse the js maximum epoch stored as a string',
-      mockValue: '8640000000000000',
+      mockValue: '8639999996400000',
```
[the remark this answers is not in the document]
Comment (MarRobSoftwire) [reply azure-functions/acbs-function/helpers/date.format-date.test.js:63]: I couldn't work it out - It seemed to only go awry when running in jest
`formatDate(8640000000000000)` was returning `275760-01-01`.
I should change the name of the test though

[the remark this answers is not in the document]
Comment (MarRobSoftwire) [reply azure-functions/acbs-function/helpers/date.format-date.test.js:63]: See test run here https://github.com/UK-Export-Finance/dtfs2/actions/runs/8175126225/job/22351558042

File: azure-functions/acbs-function/helpers/date.get-year-and-mmdd.test.js [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+    },
+    {
+      description: 'should not parse a number bigger than maximum epoch stored as a string',
+      mockValue: '8640000000000001',
+      expected: invalidYearAndMMdd,
+    },
+    {
+      description: 'should parse 0 as current date',
+      mockValue: 0,
+      expected: currentYearAndMMdd,
```
[the remark this answers is not in the document]
Comment (MarRobSoftwire) [reply azure-functions/acbs-function/helpers/date.get-year-and-mmdd.test.js:103]: It's because 0 is falsy & the old implementation used a simple `if(issedDate)`

File: azure-functions/acbs-function/helpers/date.js [hunk 6]
```diff
@@ -1,17 +1,32 @@
-const moment = require('moment');
+const { isValid, parse, format, startOfDay, add, getDaysInMonth, formatISO, differenceInMonths } = require('date-fns');
+
+const validDateFormats = [
+  'MM-dd-yy',
```
[the remark this answers is not in the document]
Comment (MarRobSoftwire) [reply azure-functions/acbs-function/helpers/date.js:4]: There were what moment was accepting - it's probably safe to remove the MM-dd-yyyy ones but it was easy enough to implement them first & ask questions later

The `yyyyyy` cases were to allow years bigger than 9999, again easy enough to remove them & change the tests if you think thats worthwhile?

File: azure-functions/acbs-function/helpers/date.format-date.test.js [hunk 7]
```diff
@@ -95,11 +88,6 @@ describe('formatDate', () => {
       mockValue: '170895577.7575',
       expected: 'Invalid date',
     },
-    {
-      description: 'should not parse a string written in scientific notation',
-      mockValue: '5e5',
```
[the remark this answers is not in the document]
Comment (MarRobSoftwire) [reply azure-functions/acbs-function/helpers/date.format-date.test.js:100]: Yeah because calling `Number('5e5')` works - it's enough of an edge case that it shouldn't cause any trouble and I can probably just delete the test

File: azure-functions/acbs-function/helpers/date.is-epoch.test.js [hunk 8]
```diff
@@ -72,16 +67,6 @@ describe('isEpoch', () => {
...
-      description: 'should return false when input is more than the Date object maximum value',
-      mockValue: 8640000000000001,
-      expected: false,
-    },
-    {
-      description: 'should return false when input is more than Date object maximum value stored as string',
```
[the remark this answers is not in the document]
Comment (MarRobSoftwire) [reply azure-functions/acbs-function/helpers/date.is-epoch.test.js:81]: I changed the implementation to just check its an integer - I'm happy to add in `<= 864000000000000` and similar if you think worthwhile?
I made the decision that these edge cases aren't actually going to impact the functionality and probably dont need to be replicated exactly

File: azure-functions/acbs-function/test-helpers/date-formats.js [hunk 9]
```diff
@@ -1,9 +1,3 @@
-const invalidDateFormats = [
```
[the remark this answers is not in the document]
Comment (MarRobSoftwire) [reply azure-functions/acbs-function/test-helpers/date-formats.js:1]: Actually it's because `date-fns` will parse the date `24/03/07` as the year 24AD

**Review by AlexBramhill — APPROVED** [review]

**Review by AlexBramhill — DISMISSED** [review]

**Review by ChristianMcc — APPROVED** [review]
smashing
